### ISSUE TITLE
Ink 4.4.0 doesn't wait for user input from `stdin` on second mount

### ISSUE THREAD

**Issue by threepointone** [vadimdemedes/ink#625]
Hiya! Looks like this commit https://github.com/vadimdemedes/ink/pull/616 broke something with input handling in 4.4.0. I managed to make a reproduction here https://github.com/threepointone/ink-repro. By running it, you'll see that hitting enter once will log the next message, but quit immediately after (with exit code 0). 

With 4.3.1, it'll wait for 3 presses, as expected.

**Comment by vadimdemedes** [comment]
Thanks for the exemplary bug report, Sunil!

I traced the issue down to https://github.com/vadimdemedes/ink/pull/616/files#diff-86963609bec6fda3d34233676d5ae33f6aa4b9bbe2758a10cd4863b704b37093R178, which gets called during `unmount()`. The way I understand it, is it basically tells Node.js "I don't need data from `process.stdin` anymore, feel free to quit if there's no I/O happening".

I verified this by adding a `process.stdin.ref()` (the opposite of `unref()`) after the first `waitForPress()` in your reproduction → https://github.com/threepointone/ink-repro/blob/main/src/index.tsx#L27.

@matteodepalo Do we have to `unref()` the `process.stdin` in Ink? Can we remove it?

**Comment by vadimdemedes** [comment]
@matteodepalo Just saw your PR (https://github.com/vadimdemedes/ink/pull/622), thanks for the fix! Could you elaborate why we need to `unref()` in the first place? I'm wondering if it's going to have other side effects, in case other code outside of Ink depends on `process.stdin` data.

**Comment by vadimdemedes** [comment]
@threepointone Just published a 4.4.1 with @matteodepalo's fix while we're figuring this stuff out.

**Comment by threepointone** [comment]
Amazing thank you! I’ll try this later when I’m at my laptop. Thanks for the quick turnaround!

**Comment by threepointone** [comment]
4.4.1 appears to have fixed my specific issue, thank you!

**Comment by threepointone** [comment]
Looks like it broke ink-testing-library tests for components that use useInput, I put up a PR with a fix https://github.com/vadimdemedes/ink-testing-library/pull/23

**Comment by matteodepalo** [comment]
>Could you elaborate why we need to unref() in the first place? I'm wondering if it's going to have other side effects, in case other code outside of Ink depends on process.stdin data.

Basically, without `unref()` all the `useInput` tests would timeout because the node process wasn't allowed to exit. This because adding any event listener will keep the process open, even if you call `.off`. This wasn't just in tests, our CLI commands wouldn't end "naturally" after everything was processed.

Before we had `pause` which would accomplish the same goal as `unref`, making sure the current process wouldn't keep listening to stdin.

Thank you so much for merging!

**Comment by matteodepalo** [comment]
To me that change isn't breaking code that uses the `ink` library itself because it changes the way an internal event is managed. The breaking change in `ink-testing-library` is because the `Stdin` class emits a `data` event manually, but I don't think that is part of the public API. If you stick to what is documented in the readme then you shouldn't get any breaking changes.

If you don't think that's enough then yes I agree that we should revert the change in a patch and ship the `readable` work as part of version 5.

**Comment by matteodepalo** [comment]
> Although I was thinking that [useStdin](https://github.com/vadimdemedes/ink#usestdin) users might see this as a breaking change, since stdin would emit data events.

Yeah I understand this. It mostly depends on what you consider to be public and private APIs. `data` events are not documented so I assumed it could be treated as private.

**Comment by matteodepalo** [comment]
Also, according to the [docs](https://nodejs.org/api/stream.html#event-data) the `data` event will still be emitted when we call `stream.read()`, so anyone subscribing to that event won't have any breaking change. The problem ink ink-testing-library was the other way around, we needed to trigger the `useInput` callback and the only way to do that is to trigger the `readable` event.

**Comment by threepointone** [comment]
Should I send a PR to ink-testing-lbrary that brings back `emit('data',...`?

**Comment by ohana54** [comment]
It broke our tests when we tried to upgrade (we use the testing library for our low level components).
I went now and manually tested one of our product flows, user input works fine AFAICT.

**Comment by sindresorhus** [comment]
This was fixed in v4.4.1 (PR #622) by calling stdin.ref() when enabling raw mode, so re-mounting correctly keeps the event loop alive.
